### PR TITLE
Feature: Factory Quotas

### PR DESCRIPTION
**Pull request by hihoman23**
### Work done
This widget (combined with a gadget) lets you set certain quotas for each unit type in the factory build menu, which are unit minimums. This means that the factory will queue new units of this type until the quota is fulfilled. These quotas are per-factory. The way you can set the quotas is by toggling a toggle, that tells the build menu to set quotas instead of setting the queue. The toggle will reset to setting the queue every time the selection is changed. The widget will use alt to queue the units.

Priority is handled like this (low numbers = high priority):
1. alt-queued units (including ones from the quota widget)
2. units from factory quotas
   a. quotas that have the most units missing from the quota (in percentage)
   b. quotas that have the least units missing from the quota
3. normally queued units

The toggle is bound to `factoryqueuemode`, and stop production will also cancel quotas.

#### Test steps
- Stop production changes:
   - [ ] Confirm that nothing breaks after queueing and other edge cases with stop production.
- Quota logic:
   - [ ] Confirm that quotas insert commands well in both normal and repeat queues
   - [ ] Set quotas to less than 0 
- Build menu interface:
   - [ ] Confirm that the build menu displays quotas like this: `actual units/quota`
   - [ ] Confirm that keyboard interaction with gridmenu works well for all keys and doesn't do anything else but set the quota
   - [ ] Confirm that both buildmenus display the text well for all text lengths.

### CODE REVIEW

**Comment by sprunk** [comment]
you don't need a gadget just to add the command, can do it widget-side

File: luaui/Widgets/unit_factory_quota.lua [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    end
+    return best, bestKey
+end
+
+---length of all tables
+---@param table table
+---@return integer
+function table.length(table)
+    local x = 0
+    for _ in pairs(table or {}) do
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:77]:
```suggestion
    for _ in pairs(table) do
```
Nil should not be treated as an empty table, this will swallow errors from passing nil.

File: luaui/Widgets/unit_factory_quota.lua [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        end
+    end
+    return best, bestKey
+end
+
+---length of all tables
+---@param table table
+---@return integer
+function table.length(table)
+    local x = 0
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:76]:
```suggestion
    local length = 0
```
Variables should use descriptive names

File: luaui/Widgets/unit_factory_quota.lua [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+-----
+
+
+
+
+--------- quota logic -------------
+---@param table table
+---@param f function|nil
+local function findMin(table, f)
+    f = f or function(i1, i2, _, _) return i1 < i2 end
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:57]: Why can a comparison function be passed as a parameter? How is minimum value applicable to a table with non-numeric values?

File: luaui/Widgets/unit_factory_quota.lua [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    local best
+    local bestKey
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:59]:
```suggestion
    local minimumValue
    local minimumKey
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+function table.length(table)
+    local x = 0
+    for _ in pairs(table or {}) do
+        x = x + 1
+    end
+    return x
+end
+
+local function isFactoryUsable(factoryID)
+    local commandq = GetFactoryCommands(factoryID, 2)
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:84]:
```suggestion
    local commandQueue = GetFactoryCommands(factoryID, 2)
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    return commandq and( #commandq == 0 or not (commandq[1].options.alt or (commandq[2] and commandq[2].options.alt)))
+end
+
+local function orderDequeue(unitID, buildDefID, count)
+	while count > 0 do
+		count = count - 100
+
+		GiveOrderToUnit(unitID, -buildDefID, {}, { "right", "ctrl", "shift" })
+	end
+end
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:97]: Can a build queue order for an arbitrary amount not be passed?

File: luaui/Widgets/unit_factory_quota.lua [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+local function clearFactQ(factID)
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:99]:
```suggestion
local function clearFactoryQueue(factID)
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+	if queue ~= nil then
+		for _, buildPair in ipairs(queue) do
+			local buildUnitDefID, count = next(buildPair, nil)
+
+			orderDequeue(unitID, buildUnitDefID, count)
+		end
+	end
+end
+
+local function putInFrontOfFactQ(...)
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:112]:
```suggestion
local function prependToFactoryQueue(...)
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    local factID, cmdID, params, opts = ...
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:113]: Isn't `opts` a table? Does this function actually need to support variable number of parameters?

Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:113]:
```suggestion
    local factIoryD, cmdID, params, opts = ...
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+        local cmd = altCmds[i]
+        GiveOrderToUnit(factID, cmd.id, cmd.params, cmd.options)
+    end
+    GiveOrderToUnit(...)
+    for _, cmd in ipairs(others) do
+        GiveOrderToUnit(factID, cmd.id, cmd.params, cmd.options)
+    end
+end
+
+local function insertToFactQ(...)
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:137]:
```suggestion
local function appendToFactoryQueue(...)
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    local factID, cmdID, params, opts = ...
+    local currCmd, targetID = Spring.GetUnitWorkerTask(factID)
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:139]:
```suggestion
    local currentCmd, targetID = Spring.GetUnitWorkerTask(factID)
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        GiveOrderToUnit(factID, cmd.id, cmd.params, cmd.options)
+    end
+    GiveOrderToUnit(...)
+    for _, cmd in ipairs(others) do
+        GiveOrderToUnit(factID, cmd.id, cmd.params, cmd.options)
+    end
+end
+
+local function insertToFactQ(...)
+    local factID, cmdID, params, opts = ...
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:138]:
```suggestion
    local factoryID, cmdID, params, opts = ...
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+    local currCmd, targetID = Spring.GetUnitWorkerTask(factID)
+    local insertnormally = true
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:140]:
```suggestion
    local insertNormally = true
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+    for factID, quota in pairs(quotas) do
+        if isFactoryUsable(factID) then
+            for udefid, num in pairs(quota) do
+                if num == 0 then
+                    quota[udefid] = nil
+                end
+            end
+            if table.length(quota)>0 then
+                local function isBetter(q1, q2, k1, k2)
+                    return (table.length((builtUnits[factID] or {})[k1] or {})/q1) < (table.length((builtUnits[factID] or {})[k2] or {})/q2)
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:164]: Can this be made more readable?

File: luaui/Widgets/unit_factory_quota.lua [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+local function fillQuotas()
+    for factID, quota in pairs(quotas) do
+        if isFactoryUsable(factID) then
+            for udefid, num in pairs(quota) do
+                if num == 0 then
+                    quota[udefid] = nil
+                end
+            end
+            if table.length(quota)>0 then
+                local function isBetter(q1, q2, k1, k2)
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:163]: What does `isBetter` mean? What does this function do? What do the parameters represent? Please give them descriptive names.

File: luaui/Widgets/unit_factory_quota.lua [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+                if quotaNum > table.length((builtUnits[factID] or {})[uDefID] or {}) then
+                    insertToFactQ(factID, -uDefID, {}, {"alt"})
+                end
+            end
+        end
+    end
+end
+
+function widget:GameFrame(n)
+    if n % 4 == 0 then -- improve perfomance
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:176]: Does this really need to run 7.5 times a second?

File: luaui/Widgets/unit_factory_quota.lua [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+        fillQuotas()
+    end
+end
+
+----- handle resetting queue mode to normal
+local function isOnQuotaBuildMode(unitID)
+	return GetUnitCmdDescs(unitID)[FindUnitCmdDesc(unitID, CMD_QUOTA_BUILD_TOGGLE)].params[1]+0 == 1
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:183]: This same logic appears in the build menu widgets, is there a single, common location this logic can be extracted to? Perhaps in the `WG['Quotas']` table.

File: luaui/Widgets/unit_factory_quota.lua [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+end
+
+function widget:SelectionChanged(newSel)
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:186]:
```suggestion
function widget:SelectionChanged(newSelection)
```

File: luaui/Widgets/unit_factory_quota.lua [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+    myTeam = Spring.GetMyTeamID()
+end
+
+function widget:Initialize()
+    widget:PlayerChanged()
+
+    for unitName, _ in pairs(quotas) do
+        quotas[UnitDefNames[unitName].id] = quotas[unitName]
+        quotas[unitName] = nil
+    end
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:236]: What is the purpose of this? It appears to be mutating the table to use unit ID's instead of names? Why not initialize the table with the desired keys?

File: luaui/Widgets/unit_factory_quota.lua [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+
+    WG.Quotas = {}
+    WG.Quotas.getQuotas = function()
+        return quotas
+    end
+    WG.Quotas.update = function(newQuotas)
+        quotas = newQuotas
+    end
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:244]: Wouldn't it be more efficient to have the update function be passed a queue ID and update the single queue, rather than having to update the entire table every time?

File: luaui/Widgets/unit_factory_quota.lua
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:1]: What happens if a factory with a quota changes hands, both to an ally or to an enemy?

**Review by WatchTheFort — CHANGES_REQUESTED** [review]

File: luarules/gadgets/unit_factory_quota.lua [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+		factoryQuotaCmdDesc.params[1] = 0
+		Spring.InsertUnitCmdDesc(unitID, factoryQuotaCmdDesc)
+	end
+end
+
+function gadget:Initialize()
+	for _, unitID in ipairs(Spring.GetAllUnits()) do
+		gadget:UnitCreated(unitID, Spring.GetUnitDefID(unitID))
+	end
+end
```
[the remark this answers is not in the document]
Comment (sprunk) [reply luarules/gadgets/unit_factory_quota.lua:57]: /luarules reload

File: luaui/Widgets/gui_buildmenu.lua [hunk 22]
```diff
@@ -984,6 +1025,50 @@ local function setPreGamestartDefID(uDefID)
 	end
 end
 
+local function isOnQuotaBuildMode(factID, targetDefID)
+	if factID then
+		return Spring.GetUnitCmdDescs(factID)[Spring.FindUnitCmdDesc(factID, CMD_QUOTA_BUILD_TOGGLE)].params[1]+0 == 1
```
[the remark this answers is not in the document]
Comment (hihoman23) [reply luaui/Widgets/gui_buildmenu.lua:1030]: Yes, the result was actually it being a string in my tests, so this is to make it an int

File: luaui/Widgets/unit_factory_quota.lua [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+      license = "GNU GPL, v2 or later",
+      layer = -1,
+      enabled = true
+    }
+end
+
+VFS.Include('luarules/configs/customcmds.h.lua')
+
+local maxBuildProg = 0.075 -- maximum build progress that gets replaced in a repeat queue
+local maxMetal = 500 -- maximum metal cost that gets replaced in a repeat queue(7.5% of a juggernaut is still over 2k metal)
```
[the remark this answers is not in the document]
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:17]: This is new behavior, to replace needing to check every frame to insert.

File: luaui/Widgets/unit_factory_quota.lua (hunk 19 above)
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:236]: The purpose of this was for testing reasons (it allowed me to hardcode a quota in the code). Thank you for noticing.

File: luaui/Widgets/unit_factory_quota.lua (hunk 20 above)
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:244]: I don't even know if this is necessary, since the table is never copied in the other widgets, so I even think calling this might be unnecessary.

**Comment by hihoman23** [comment]
> you don't need a gadget just to add the command, can do it widget-side

I think this only applies to commands, but not toggles.

File: luaui/Widgets/unit_factory_quota.lua
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:1]: It gets ignored for the rest of the game.

File: luaui/Widgets/unit_factory_quota.lua (hunk 20 above)
Comment (przystuj) [reply luaui/Widgets/unit_factory_quota.lua:244]: I use this update in buildings_and_units_tracker, pls don't remove :D

Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:244]: > I use this update in buildings_and_units_tracker, pls don't remove :D

```lua
local quotas = WG.Quotas.getQuotas()
quotas = changedQuotas
```
This should do the same thing as `WG.Quotas.update(changedQuotas)`

File: luaui/Widgets/unit_factory_quota.lua [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+---length of all tables
+---@param table table
+---@return integer
+function table.length(table)
+    local x = 0
+    for _ in pairs(table or {}) do
+        x = x + 1
+    end
+    return x
+end
```
[the remark this answers is not in the document]
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua]: I found out that `table.count` exists.

File: luaui/Widgets/unit_factory_quota.lua (hunk 23 above)
[the remark this answers is not in the document]
Comment (WatchTheFort) [reply luaui/Widgets/unit_factory_quota.lua:18]: I meant basing the decision on metal value spent rather than build progress percent

File: luaui/Widgets/unit_factory_quota.lua
Comment (WatchTheFort) [reply luaui/Widgets/unit_factory_quota.lua:1]: As in, any existing quotas for that factory get cleared?

File: luaui/Widgets/unit_factory_quota.lua [hunk 25]
```diff
@@ -31,141 +32,138 @@ for unitDefID, uDef in pairs(UnitDefs) do
...
+    local minimumUnitDefID
+    local minimumRatio
+    for unitDefID, quotaNumber in pairs(quota) do
+        if not minimumQuota then
+            minimumQuota = quotaNumber
+            minimumUnitDefID = unitDefID
+            minimumRatio = getNumberOfUnits(factoryID, unitDefID)/quotaNumber
+            
+        else
+            local currentRatio = getNumberOfUnits(factoryID, unitDefID)/quotaNumber
```
Comment (WatchTheFort) [inline luaui/Widgets/unit_factory_quota.lua:73]: Is `quotaNumber` guaranteed to never be 0?

Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:73]: Yes, they are filtered out in lines 158 - 162

File: luaui/Widgets/unit_factory_quota.lua
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:1]: I'm sorry, I misread your question. Units that are shared get cleared from the unit count, but factories actually keep their quotas, which is an oversight of mine. The question that remains is if this is wanted behavior.

Comment (WatchTheFort) [reply luaui/Widgets/unit_factory_quota.lua:1]: For an enemy, a captured factory should clear all orders. For an ally, what happens right now when transferring a factory, does it keep its queue?

File: luaui/Widgets/unit_factory_quota.lua (hunk 23 above)
[the remark this answers is not in the document]
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:18]: It needs to be under both values to pass through. I don't know of any existing behavior that takes build progress and metal value into consideration when canceling commands.

File: luaui/Widgets/unit_factory_quota.lua (hunk 20 above)
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:236]: Would it be possible if I remove it from all other widgets, but keep it exposed for backwards compatibility?

File: luaui/Widgets/unit_factory_quota.lua
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua:1]: It also gets canceled.

Comment (WatchTheFort) [reply luaui/Widgets/unit_factory_quota.lua:1]: In that case I would keep the behaviour consistent and also cancel any quotas

File: luaui/Widgets/unit_factory_quota.lua (hunk 20 above)
Comment (WatchTheFort) [reply luaui/Widgets/unit_factory_quota.lua:236]: > I use this update in buildings_and_units_tracker, pls don't remove :D

How could you be using this functionality when it's not merged yet?

File: luaui/Widgets/unit_factory_quota.lua (hunk 6 above)
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua]: I managed to remove that entire part in the current update.

File: luaui/Widgets/unit_factory_quota.lua (hunk 16 above)
Comment (hihoman23) [reply luaui/Widgets/unit_factory_quota.lua]: No, it's changed to 2 times a second now.

File: luaui/Widgets/gui_gridmenu.lua [hunk 26]
```diff
@@ -2528,8 +2550,8 @@ function widget:CommandNotify(cmdID, _, cmdOpts)
 	end
 end
 
--- widget:UnitCommand(unitID, unitDefID, unitTeam, cmdID, cmdTag, cmdParams, cmdOpts)
-function widget:UnitCommand(unitID, _, _, cmdID, _, cmdParams)
+-- widget:UnitCommand(unitID, unitDefID, unitTeam, cmdParams, cmdOpts, cmdID, cmdTag)
```
Comment (WatchTheFort) [inline luaui/Widgets/gui_gridmenu.lua:2553]: What is the purpose of this comment?

**Review by WatchTheFort — CHANGES_REQUESTED** [review]

File: luaui/Widgets/gui_gridmenu.lua (hunk 26 above)
Comment (hihoman23) [reply luaui/Widgets/gui_gridmenu.lua:2553]: It's there to show what was removed using `_` and it was also there before, so I decided to update it (since it was wrong)

Comment (WatchTheFort) [reply luaui/Widgets/gui_gridmenu.lua:2553]: Just put them in the function parameters and remove the comment, it's fine for engine function signatures to have the unused parameters in them.

**Review by WatchTheFort — APPROVED** [review]

**Comment by hihoman23** [comment]
Note before merging: please note this side effect from returning true on AllowCommand for stopproduction, which is that if the keybind is put in front of stop in uikeys.txt, it will now execute both commands.

**Comment by icexuick** [comment]
Wouldn't you want a quota as in a "percentage" of units, but continuously producing them?
Maybe this is more meant to have at least 5 T1 cons and 15 rezbots out on the field, and use the regular queuing to keep producing your army.

But I could also love the idea of wanting a mixed composition of your force being continuously be build. Like:
- 25% snipers
- 50% welders
- 5% rezbots
- 20% AA

The position on the quota in the build-menu is pretty nice, though it still obscuring things (and removing showing the cost.
I'm unsure how to solve it better though - maybe with new RMLUI we can use smaller fonts to fit it all better.

Other than ^^ I can't comment more on usability/gameplay aspects. I'm a too casual player for this and I don't expect/see me using this.

**Comment by SethDGamre** [comment]
> Wouldn't you want a quota as in a "percentage" of units, but continuously producing them?  
> But I could also love the idea of wanting a mixed composition of your force being continuously be build. Like:
> 
> * 25% snipers
> * 50% welders
> * 5% rezbots
> * 20% AA

This would be worse. If all I wanted was five of a specific type of unit on the field, I would have to do percentage math in my brain to have a specific number of that unit in existence. 

Also, maybe I only want to have exactly three radar bots at any given time because I only have three armies. Having a fourth and a fifth doesn't help me and it's just a waste.

Maybe just because I have 100 tigers and 25 banishers doesn't mean that I want a proportional 10% resbots or jammers. I would have to adjust my percentages constantly to account for my army size so I wouldn't overproduce unnecessary units.

I also can't imagine a way to communicate the percentages to the player in a way that's intuitive.

The current iteration of factory quotas is intuitive and works really well imo.

**Comment by hihoman23** [comment]
The [stopproduction side effect](https://github.com/beyond-all-reason/Beyond-All-Reason/pull/3688#issuecomment-2357551586) is now fixed

File: luarules/gadgets/cmd_factory_stop_production.lua [hunk 27]
```diff
@@ -15,21 +15,7 @@ end
...
---------------------------------------------------------------------------------
-
-local spGetRealBuildQueue = Spring.GetRealBuildQueue
-local spGiveOrderToUnit = Spring.GiveOrderToUnit
-local spInsertUnitCmdDesc = Spring.InsertUnitCmdDesc
-
-local CMD_WAIT = CMD.WAIT
-local EMPTY = {}
-local DEQUEUE_OPTS = { "right", "ctrl", "shift" } -- right: dequeue, ctrl+shift: 100
+local magicValue = "StopProduction"
```
Comment (WatchTheFort) [inline luarules/gadgets/cmd_factory_stop_production.lua:18]: Is there a more descriptive name for this variable? What does the value represent, the name of the command? Why not use the CMD table and [Spring.GetUnitCmdDescs](https://springrts.com/wiki/Lua_SyncedRead#GetUnitCmdDescs)?

File: luarules/gadgets/cmd_factory_stop_production.lua [hunk 28]
```diff
@@ -41,86 +27,121 @@ for udid = 1, #UnitDefs do
...
+		gadgetHandler:RegisterCMDID(CMD_STOP_PRODUCTION)
+		for _, unitID in pairs(Spring.GetAllUnits()) do
+			gadget:UnitCreated(unitID, Spring.GetUnitDefID(unitID))
+		end
+	end
+else
+	local scriptUnitCommand = Script.LuaUI.UnitCommand
+
+	local spGetSpectatingState = Spring.GetSpectatingState
+	local spAreTeamsAllied = Spring.AreTeamsAllied
```
Comment (WatchTheFort) [inline luarules/gadgets/cmd_factory_stop_production.lua:128]: Do these actually need to be localized? Are they being called multiple times per frame?

File: luarules/gadgets/cmd_factory_stop_production.lua (hunk 27 above)
Comment (hihoman23) [reply luarules/gadgets/cmd_factory_stop_production.lua:18]: The value represents the first argument passed on to `SendToUnsynced` that is usually used as an identifier so the correct gadget receives the information.

Comment (WatchTheFort) [reply luarules/gadgets/cmd_factory_stop_production.lua:18]:
```suggestion
local identifier = "StopProduction"
```

**Review by WatchTheFort — APPROVED** [review]

**Comment by WatchTheFort** [comment]
@hihoman23 Had to revert because it was causing errors when selecting anything in a mobile builder's build menu.

```txt
[t=00:08:48.504481][f=0000385] Error in MousePress(): [string "LuaUI/Widgets/unit_factory_quota.lua"]:136: attempt to index field '?' (a nil value)
[t=00:08:48.504512][f=0000385] Removed widget: Grid menu
```

**Comment by hihoman23** [comment]
This is fixed and the PR is reopened in #3886
